File: z0/query.zig
//! query.zig — Z0 offline reverse geocoder, Zig runtime.
//!
//! Reads an RGEO0002 binary blob (produced by builder.zig) and answers
//! reverse-geocode queries with no heap allocations.
//!
//! API (library):
//!   const gc = try ReverseGeocoder.init(data_slice);
//!   const admin_id = gc.lookup(lat, lon);   // ?u16, null = ocean
//!
//! WASM exports (compiled with -target wasm32-freestanding):
//!   geocoder_init(ptr, len) i32   — 0 on success, -1 on bad file
//!   geocoder_lookup(lat, lon) i32 — admin_id (0–65533) or -1 (ocean)

const std = @import("std");

// ---------------------------------------------------------------------------
// Constants (must match builder.zig exactly)
// ---------------------------------------------------------------------------

const MAGIC           = "RGEO0002";
const GRID_COLS: usize = 1440;
const GRID_ROWS: usize = 720;
const GRID_CELL_DEG: f64 = 0.25;
const SENTINEL_BOUNDARY: u16 = 0xFFFF;
const SENTINEL_OCEAN:    u16 = 0xFFFE;
const BLOCK_SIZE:    usize = 64;
const BLOCK_RECORDS: usize = 10;
const RECORD_SIZE:   usize = 6;   // u32 morton + u16 admin_id
const MORTON_STEPS:  f64   = 4096.0;

// ---------------------------------------------------------------------------
// Morton code
// ---------------------------------------------------------------------------

fn mortonSpread(v: u32) u32 {
    var x: u32 = v & 0xFFF;
    x = (x | (x << 8)) & 0x00FF00FF;
    x = (x | (x << 4)) & 0x0F0F0F0F;
    x = (x | (x << 2)) & 0x33333333;
    x = (x | (x << 1)) & 0x55555555;
    return x;
}

fn mortonEncode(lat: f64, lon: f64) u32 {
    const lq: u32 = @intFromFloat(@min(MORTON_STEPS - 1, @max(0.0, (lat + 90.0)  / 180.0 * MORTON_STEPS)));
    const aq: u32 = @intFromFloat(@min(MORTON_STEPS - 1, @max(0.0, (lon + 180.0) / 360.0 * MORTON_STEPS)));
    return mortonSpread(lq) | (mortonSpread(aq) << 1);
}

// ---------------------------------------------------------------------------
// Read helpers (little-endian, unaligned)
// ---------------------------------------------------------------------------

inline fn rdU16(data: []const u8, off: usize) u16 {
    return std.mem.readInt(u16, data[off..][0..2], .little);
}
inline fn rdU32(data: []const u8, off: usize) u32 {
    return std.mem.readInt(u32, data[off..][0..4], .little);
}

// ---------------------------------------------------------------------------
// ReverseGeocoder
// ---------------------------------------------------------------------------

pub const ReverseGeocoder = struct {
    data: []const u8,

    bitmap_off:          usize,
    rank_off:            usize,
    values_off:          usize,
    land_cell_count:     usize,
    morton_block_off:    usize,
    morton_dir_off:      usize,
    morton_record_count: usize,
    morton_block_count:  usize,
    admin_off:           usize,

    pub const Error = error{ BadMagic, BadVersion };

    /// Initialise from a raw RGEO0002 binary slice (no allocation).
    pub fn init(data: []const u8) Error!ReverseGeocoder {
        if (data.len < 64) return Error.BadMagic;
        if (!std.mem.eql(u8, data[0..8], MAGIC)) return Error.BadMagic;
        const version = rdU32(data, 8);
        if (version != 1) return Error.BadVersion;

        // Header layout (see builder.zig writeBinaryFile):
        //  [0:8]  magic        [8:12]  version       [12:16] timestamp
        //  [16:20] bitmap_off  [20:24] rank_off       [24:28] values_off
        //  [28:32] land_cells  [32:36] morton_blk_off [36:40] morton_dir_off
        //  [40:44] m_rec_count [44:48] m_blk_count    [48:52] admin_off
        //  [52:56] name_off    [56:64] reserved
        return .{
            .data                = data,
            .bitmap_off          = rdU32(data, 16),
            .rank_off            = rdU32(data, 20),
            .values_off          = rdU32(data, 24),
            .land_cell_count     = rdU32(data, 28),
            .morton_block_off    = rdU32(data, 32),
            .morton_dir_off      = rdU32(data, 36),
            .morton_record_count = rdU32(data, 40),
            .morton_block_count  = rdU32(data, 44),
            .admin_off           = rdU32(data, 48),
        };
    }

    /// Reverse-geocode (lat, lon). Returns admin_id or null for ocean.
    pub fn lookup(self: *const ReverseGeocoder, lat: f64, lon: f64) ?u16 {
        const la = std.math.clamp(lat, -90.0,  90.0);
        const lo = std.math.clamp(lon, -180.0, 180.0);

        // Layer 0: 0.25° coarse grid
        const col: usize = @min(GRID_COLS - 1,
            @as(usize, @intFromFloat(@max(0.0, (lo + 180.0) / GRID_CELL_DEG))));
        const row: usize = @min(GRID_ROWS - 1,
            @as(usize, @intFromFloat(@max(0.0, (90.0 - la) / GRID_CELL_DEG))));
        const idx = row * GRID_COLS + col;

        // Bitmap test
        const byte_idx = self.bitmap_off + idx / 8;
        if ((self.data[byte_idx] >> @intCast(idx & 7)) & 1 == 0) return null;

        // Rank → values[]
        const rank  = self.bitmapRank(idx);
        const value = rdU16(self.data, self.values_off + rank * 2);

        if (value == SENTINEL_OCEAN)    return null;
        if (value != SENTINEL_BOUNDARY) return value;  // interior fast path

        // Layer 1: Morton boundary table
        const m = mortonEncode(la, lo);
        if (self.blockSearch(m))        |aid| return aid;
        if (self.blockSearch(m >> 2))   |aid| return aid;
        return null;
    }

    // -- Bitmap rank (O(1) using file's precomputed rank table) --------------

    fn bitmapRank(self: *const ReverseGeocoder, idx: usize) usize {
        const block_idx  = idx / 512;
        var rank: usize  = rdU32(self.data, self.rank_off + block_idx * 4);

        const block_byte = self.bitmap_off + block_idx * 64;  // 512 bits / 8
        const local_bit  = idx % 512;
        const full_bytes = local_bit / 8;
        const rem_bits:u3 = @intCast(local_bit % 8);

        for (0..full_bytes) |i| {
            rank += @popCount(self.data[block_byte + i]);
        }
        if (rem_bits > 0) {
            const mask: u8 = (@as(u8, 1) << rem_bits) - 1;
            rank += @popCount(self.data[block_byte + full_bytes] & mask);
        }
        return rank;
    }

    // -- Block binary search --------------------------------------------------

    fn blockSearch(self: *const ReverseGeocoder, m: u32) ?u16 {
        const n = self.morton_block_count;
        if (n == 0) return null;

        // upper_bound: find first block where dir[block] > m, then go back one
        var lo: usize = 0;
        var hi: usize = n;
        while (lo < hi) {
            const mid = lo + (hi - lo) / 2;
            if (rdU32(self.data, self.morton_dir_off + mid * 4) <= m) {
                lo = mid + 1;
            } else {
                hi = mid;
            }
        }
        if (lo == 0) return null;  // all blocks start above m
        const block_idx = lo - 1;

        const block_start = self.morton_block_off + block_idx * BLOCK_SIZE;
        for (0..BLOCK_RECORDS) |i| {
            const rec_off = block_start + i * RECORD_SIZE;
            const rec_m   = rdU32(self.data, rec_off);
            if (rec_m > m) break;
            if (rec_m == m) return rdU16(self.data, rec_off + 4);
        }
        return null;
    }
};

// ---------------------------------------------------------------------------
// WASM / flat-C exports
// ---------------------------------------------------------------------------

var _geocoder: ?ReverseGeocoder = null;

pub export fn geocoder_init(ptr: usize, len: usize) i32 {
    const data: [*]const u8 = @ptrFromInt(ptr);
    _geocoder = ReverseGeocoder.init(data[0..len]) catch return -1;
    return 0;
}

/// Returns admin_id (0–65533) or -1 for ocean / error.
pub export fn geocoder_lookup(lat: f64, lon: f64) i32 {
    const gc = _geocoder orelse return -1;
    return if (gc.lookup(lat, lon)) |id| @intCast(id) else -1;
}

// ---------------------------------------------------------------------------
// Native CLI entry point
// ---------------------------------------------------------------------------

pub fn main() !void {
    const args = try std.process.argsAlloc(std.heap.page_allocator);
    defer std.process.argsFree(std.heap.page_allocator, args);

    if (args.len < 3) {
        std.debug.print("Usage: query <lat> <lon> [z0_geo.bin]\n", .{});
        return;
    }
    const lat = try std.fmt.parseFloat(f64, args[1]);
    const lon = try std.fmt.parseFloat(f64, args[2]);
    const path = if (args.len >= 4) args[3] else "z0_geo.bin";

    const file = try std.fs.cwd().openFile(path, .{});
    defer file.close();
    const data = try file.readToEndAlloc(std.heap.page_allocator, 256 * 1024 * 1024);
    defer std.heap.page_allocator.free(data);

    const gc = try ReverseGeocoder.init(data);
    if (gc.lookup(lat, lon)) |id| {
        std.debug.print("admin_id: {d}\n", .{id});
    } else {
        std.debug.print("ocean / unclassified\n", .{});
    }
}

File: z0/query.py
#!/usr/bin/env python3
"""
query.py — Offline reverse geocoder query engine.

Loads z0_geo.bin and provides reverse geocode lookups.

Usage:
    python query.py <lat> <lon>
    python query.py 48.8566 2.3522    # Paris, France
"""

import bisect
import json
import mmap
import os
import struct
import sys
from typing import Optional

import zstandard as zstd


# ---------------------------------------------------------------------------
# Constants
# ---------------------------------------------------------------------------

MAGIC = b"RGEO0002"
HEADER_SIZE = 64

GRID_COLS = 1440
GRID_ROWS = 720
GRID_CELL_DEG = 0.25

SENTINEL_BOUNDARY = 0xFFFF
SENTINEL_OCEAN = 0xFFFE

BLOCK_RECORDS = 10
BLOCK_SIZE = 64
RECORD_SIZE = 6   # uint32 morton + uint16 admin_id


# ---------------------------------------------------------------------------
# Morton code (query side)
# ---------------------------------------------------------------------------

def interleave_bits(x: int, y: int) -> int:
    """Interleave bits of two 16-bit integers into a 32-bit Morton code."""
    result = 0
    for i in range(16):
        result |= ((x >> i) & 1) << (2 * i)
        result |= ((y >> i) & 1) << (2 * i + 1)
    return result


MORTON_STEPS = 4096  # 12-bit quantization; must match builder

def compute_morton(lat: float, lon: float) -> int:
    lat_q = int((lat + 90.0) / 180.0 * MORTON_STEPS) & (MORTON_STEPS - 1)
    lon_q = int((lon + 180.0) / 360.0 * MORTON_STEPS) & (MORTON_STEPS - 1)
    return interleave_bits(lat_q, lon_q)


# ---------------------------------------------------------------------------
# ReverseGeocoder
# ---------------------------------------------------------------------------

class ReverseGeocoder:
    """
    Offline reverse geocoder backed by a memory-mapped z0_geo.bin file.

    Thread-safe for concurrent reads (mmap is read-only after __init__).
    """

    def __init__(self, data_path: str = "z0_geo.bin"):
        self._f = open(data_path, "rb")
        self._mm = mmap.mmap(self._f.fileno(), 0, access=mmap.ACCESS_READ)
        self._parse_header()
        self._load_name_tables()

    def close(self):
        self._mm.close()
        self._f.close()

    def __enter__(self):
        return self

    def __exit__(self, *args):
        self.close()

    # ------------------------------------------------------------------
    # Header parsing
    # ------------------------------------------------------------------

    def _parse_header(self):
        mm = self._mm

        # Verify magic
        magic = mm[0:8]
        if magic != MAGIC:
            raise ValueError(f"Bad magic: {magic!r}, expected {MAGIC!r}")

        # Unpack header fields
        (
            _fmt_version,
            _timestamp,
            self._bitmap_offset,
            self._rank_offset,
            self._values_offset,
            self._land_cell_count,
            self._morton_block_offset,
            self._morton_dir_offset,
            self._morton_record_count,
            self._morton_block_count,
            self._admin_offset,
            self._name_offset,
            _reserved,
        ) = struct.unpack_from("<IIIIIIIIIIIIq", mm, 8)

        # Cache admin table length: each entry is 5 bytes (uint8 + uint16 + uint16)
        self._admin_entry_size = 5
        # Derive admin count from name_offset - admin_offset
        admin_section_len = self._name_offset - self._admin_offset
        self._admin_count = admin_section_len // self._admin_entry_size

        # Pre-build Morton directory as a Python list for bisect
        # Each directory entry is a uint32 (first morton code of that block)
        dir_count = self._morton_block_count
        dir_data = mm[self._morton_dir_offset: self._morton_dir_offset + dir_count * 4]
        self._directory = list(struct.unpack_from(f"<{dir_count}I", dir_data))

    # ------------------------------------------------------------------
    # Name table loading (zstd decompress once at startup)
    # ------------------------------------------------------------------

    def _load_name_tables(self):
        mm = self._mm
        name_data = bytes(mm[self._name_offset:])
        dctx = zstd.ZstdDecompressor()
        raw = dctx.decompress(name_data)
        tables = json.loads(raw.decode("utf-8"))
        self._countries = tables["countries"]
        self._adm1s = tables["adm1s"]
        self._adm2s = tables["adm2s"]

    # ------------------------------------------------------------------
    # Layer 0: coarse grid lookup
    # ------------------------------------------------------------------

    def _bitmap_bit(self, idx: int) -> bool:
        """Return True if bit `idx` is set in the bitmap."""
        byte_idx = self._bitmap_offset + idx // 8
        bit_idx = idx % 8
        return bool((self._mm[byte_idx] >> bit_idx) & 1)

    def _bitmap_rank(self, idx: int) -> int:
        """
        Return the number of set bits in the bitmap at positions 0..(idx-1).
        Uses the precomputed rank table (one entry per 512-bit block) plus
        a popcount of the partial block up to `idx`.
        """
        mm = self._mm
        block_idx = idx // 512
        # Read precomputed cumulative count before this block
        rank_entry_offset = self._rank_offset + block_idx * 4
        base_rank = struct.unpack_from("<I", mm, rank_entry_offset)[0]

        # Popcount the bits within this block from the block start to idx (exclusive)
        block_start_bit = block_idx * 512
        block_start_byte = self._bitmap_offset + block_start_bit // 8

        # Number of complete bytes before idx within this block
        bits_into_block = idx - block_start_bit
        complete_bytes = bits_into_block // 8
        remaining_bits = bits_into_block % 8

        partial = 0
        for i in range(complete_bytes):
            partial += bin(mm[block_start_byte + i]).count('1')

        if remaining_bits > 0:
            last_byte = mm[block_start_byte + complete_bytes]
            # Count only the low `remaining_bits` bits
            mask = (1 << remaining_bits) - 1
            partial += bin(last_byte & mask).count('1')

        return base_rank + partial

    def _grid_lookup(self, lat: float, lon: float) -> Optional[int]:
        """
        Layer 0 coarse grid lookup.
        Returns:
            None          — ocean (no land in this cell)
            SENTINEL_OCEAN  (0xFFFE) — ocean within a land-flagged coastal cell
            SENTINEL_BOUNDARY (0xFFFF) — boundary; caller must use Layer 1
            int (0-0xFFFD) — interior; direct admin_id
        """
        col = int((lon + 180.0) / GRID_CELL_DEG)
        row = int((90.0 - lat) / GRID_CELL_DEG)

        # Clamp to valid range
        col = max(0, min(GRID_COLS - 1, col))
        row = max(0, min(GRID_ROWS - 1, row))

        idx = row * GRID_COLS + col

        if not self._bitmap_bit(idx):
            return None  # ocean — fast path

        rank = self._bitmap_rank(idx)
        # Read value: rank-th uint16 in the values array (0-based)
        val_offset = self._values_offset + rank * 2
        value = struct.unpack_from("<H", self._mm, val_offset)[0]
        return value

    # ------------------------------------------------------------------
    # Layer 1: Morton boundary table lookup
    # ------------------------------------------------------------------

    def _block_search(self, morton: int) -> Optional[int]:
        """
        Search the Morton directory for the block that could contain `morton`,
        then linearly scan that block.
        Returns admin_id or None.
        """
        directory = self._directory
        if not directory:
            return None

        # Find the last block whose first key <= morton
        pos = bisect.bisect_right(directory, morton) - 1
        if pos < 0:
            return None

        # Linear scan within the block (at most 10 records)
        block_offset = self._morton_block_offset + pos * BLOCK_SIZE
        mm = self._mm

        for i in range(BLOCK_RECORDS):
            rec_offset = block_offset + i * RECORD_SIZE
            rec_morton, admin_id = struct.unpack_from("<IH", mm, rec_offset)

            # Records are sorted; if we've passed morton, stop
            if rec_morton > morton:
                break

            # Padding records at end of block have morton=0 (only if block not full)
            # A true zero morton (south pole, prime meridian) is valid but rare;
            # we handle it because we only stop on strictly greater.
            if rec_morton == morton:
                return admin_id

        return None

    def _morton_lookup(self, lat: float, lon: float) -> Optional[int]:
        """
        Layer 1 Morton boundary table lookup with parent fallback.
        Returns admin_id or None.
        """
        morton = compute_morton(lat, lon)
        admin_id = self._block_search(morton)
        if admin_id is not None:
            return admin_id

        # Parent fallback: mask out bottom 2 bits
        morton_parent = morton >> 2
        admin_id = self._block_search(morton_parent)
        return admin_id

    # ------------------------------------------------------------------
    # Admin table lookup
    # ------------------------------------------------------------------

    def _get_admin(self, admin_id: int) -> dict:
        """Return dict with country, adm1, adm2 strings for admin_id."""
        offset = self._admin_offset + admin_id * self._admin_entry_size
        c_idx, a1_idx, a2_idx = struct.unpack_from("<BHH", self._mm, offset)
        return {
            "country": self._countries[c_idx],
            "adm1": self._adm1s[a1_idx],
            "adm2": self._adm2s[a2_idx],
        }

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    def lookup(self, lat: float, lon: float) -> Optional[dict]:
        """
        Reverse geocode (lat, lon).

        Returns dict with keys 'country', 'adm1', 'adm2' for land points,
        or None for ocean / unclassified territory.

        Lat range: [-90, 90], Lon range: [-180, 180].
        """
        # Clamp inputs to valid range
        lat = max(-90.0, min(90.0, lat))
        lon = max(-180.0, min(180.0, lon))

        # Layer 0
        grid_value = self._grid_lookup(lat, lon)

        if grid_value is None:
            return None  # ocean

        if grid_value == SENTINEL_OCEAN:
            return None  # ocean within coastal cell

        if grid_value <= 0xFFFD:
            # Interior: fast path
            return self._get_admin(grid_value)

        # BOUNDARY (0xFFFF): fall through to Layer 1
        admin_id = self._morton_lookup(lat, lon)
        if admin_id is None:
            return None

        return self._get_admin(admin_id)


# ---------------------------------------------------------------------------
# CLI entry point
# ---------------------------------------------------------------------------

def main():
    if len(sys.argv) not in (3, 4):
        print("Usage: python query.py <lat> <lon> [data_file]", file=sys.stderr)
        print("  Example: python query.py 48.8566 2.3522", file=sys.stderr)
        sys.exit(1)

    try:
        lat = float(sys.argv[1])
        lon = float(sys.argv[2])
    except ValueError:
        print("ERROR: lat and lon must be floating-point numbers.", file=sys.stderr)
        sys.exit(1)

    data_path = sys.argv[3] if len(sys.argv) == 4 else "z0_geo.bin"

    if not os.path.exists(data_path):
        print(f"ERROR: Data file not found: {data_path}", file=sys.stderr)
        print("  Run builder.py first to generate the data file.", file=sys.stderr)
        sys.exit(1)

    with ReverseGeocoder(data_path) as rg:
        result = rg.lookup(lat, lon)

    if result is None:
        print(f"({lat}, {lon}) -> ocean / unclassified")
    else:
        print(f"({lat}, {lon}) ->")
        print(f"  country: {result['country']}")
        print(f"  adm1:    {result['adm1']}")
        print(f"  adm2:    {result['adm2']}")


if __name__ == "__main__":
    main()
